### PR TITLE
Add PyTorch with RawKernel example to docs

### LINKED ISSUE #5833: Document how to write autograd-supported PyTorch function with CuPy
It is a common use-case to define [PyTorch autograd functions](https://pytorch.org/docs/stable/notes/extending.html) with CuPy's RawKernel. It would be nice to cover that in the interoperability docs.

Schenario: define `torch.autograd.Function`, implement `forward` and `backward` using `cupy.RawKernel` (zero-copy using `cupy.from_dlpack(toDlpack(tensor.detach()))`), and use it from `torch.nn.Module`.

### PR DESCRIPTION
**Pull request by emcastillo**
Blocked by #5970 

Closes #5833

### CODE REVIEW

File: docs/source/user_guide/interoperability.rst [hunk 1]
```diff
@@ -198,6 +198,76 @@ PyTorch also supports zero-copy data exchange through ``DLPack`` (see :ref:`dlpa
...
+            # Enforce contiguous arrays to simplify RawKernel indexing.
+            cupy_x = cupy.ascontiguousarray(cupy.from_dlpack(x.detach()))
+            cupy_y = cupy.empty(cupy_x.shape, dtype=cupy_x.dtype)
+            x_size = cupy_x.size
+            bs = 128
+            cupy_custom_kernel_fwd(
+                (bs,), ((x_size + bs - 1) // bs,), (cupy_x, cupy_y, x_size)
+            )
+            torch_y = torch.from_dlpack(cupy_y)
+            return torch_y
```
Comment (leofang) [inline docs/source/user_guide/interoperability.rst:252]: Not sure if this is worth mentioning: the ownership of the device memory backing `cupy_y` is implicitly transferred to `torch_y`, so this operation is safe even after going out of scope of this function.

File: docs/source/user_guide/interoperability.rst [hunk 2]
```diff
@@ -198,6 +198,76 @@ PyTorch also supports zero-copy data exchange through ``DLPack`` (see :ref:`dlpa
...
+            cupy_grad_x = cupy.zeros(cupy_grad_y.shape, dtype=cupy_grad_y.dtype)
+            gy_size = cupy_grad_y.size
+            bs = 128
+            cupy_custom_kernel_bwd(
+                (bs,),
+                ((gy_size + bs - 1) // bs,),
+                (cupy_input, cupy_grad_y, cupy_grad_x, gy_size),
+            )
+            torch_grad_x = torch.from_dlpack(cupy_grad_x)
+            return torch_grad_x
```
Comment (leofang) [inline docs/source/user_guide/interoperability.rst:268]: ditto

File: docs/source/user_guide/interoperability.rst (hunk 1 above)
Comment (emcastillo) [reply docs/source/user_guide/interoperability.rst:255]: Added!!! thanks!

File: docs/source/user_guide/interoperability.rst [hunk 3]
```diff
@@ -198,6 +198,82 @@ PyTorch also supports zero-copy data exchange through ``DLPack`` (see :ref:`dlpa
    ...     ...
 
 
+Using custom kernels in PyTorch
+-------------------------------
+
+With the DLPack protocol, it becomes very simple to implement functions in pytorch using CuPy user-defined kernels. Below is the example of a PyTorch autograd function
+that computes the forward and backward pass of the logarithm using `cupy.RawKernel` s.
```
Comment (kmaehashi) [inline docs/source/user_guide/interoperability.rst:205]:
```suggestion
that computes the forward and backward pass of the logarithm using :class:`cupy.RawKernel` s.
```

File: docs/source/user_guide/interoperability.rst [hunk 4]
```diff
@@ -198,6 +198,82 @@ PyTorch also supports zero-copy data exchange through ``DLPack`` (see :ref:`dlpa
    ...     ...
 
 
+Using custom kernels in PyTorch
+-------------------------------
+
+With the DLPack protocol, it becomes very simple to implement functions in pytorch using CuPy user-defined kernels. Below is the example of a PyTorch autograd function
```
Comment (kmaehashi) [inline docs/source/user_guide/interoperability.rst:204]:
```suggestion
With the DLPack protocol, it becomes very simple to implement functions in PyTorch using CuPy user-defined kernels. Below is the example of a PyTorch autograd function
```

File: docs/source/user_guide/interoperability.rst [hunk 5]
```diff
@@ -198,6 +198,82 @@ PyTorch also supports zero-copy data exchange through ``DLPack`` (see :ref:`dlpa
    ...     ...
 
 
+Using custom kernels in PyTorch
+-------------------------------
```
Comment (kmaehashi) [inline docs/source/user_guide/interoperability.rst:202]:
```suggestion
~~~~~~~~~~~~~~~~~~~~~~~
```

**Review by kmaehashi — CHANGES_REQUESTED** [review]
Great! LGTM except for nits.

File: docs/source/user_guide/interoperability.rst [hunk 6]
```diff
@@ -198,6 +198,82 @@ PyTorch also supports zero-copy data exchange through ``DLPack`` (see :ref:`dlpa
...
+        "cupy_custom_kernel_bwd",
+    )
+    
+    
+    class CuPyLog(torch.autograd.Function):
+        @staticmethod
+        def forward(ctx, x):
+            ctx.input = x
+            # Enforce contiguous arrays to simplify RawKernel indexing.
+            cupy_x = cupy.ascontiguousarray(cupy.from_dlpack(x.detach()))
```
Comment (leofang) [inline docs/source/user_guide/interoperability.rst:244]: Silly question: Would there be any impact in PyTorch's workflow with `x` being detached here?

File: docs/source/user_guide/interoperability.rst [hunk 7]
```diff
@@ -198,6 +198,82 @@ PyTorch also supports zero-copy data exchange through ``DLPack`` (see :ref:`dlpa
...
+            # the ownership of the device memory backing cupy_y is implicitly
+            # transferred to torch_y, so this operation is safe even after
+            # going out of scope of this function.
+            torch_y = torch.from_dlpack(cupy_y)
+            return torch_y
+    
+        @staticmethod
+        def backward(ctx, grad_y):
+            # Enforce contiguous arrays to simplify RawKernel indexing.
+            cupy_input = cupy.from_dlpack(ctx.input.detach()).ravel()
```
Comment (leofang) [inline docs/source/user_guide/interoperability.rst:260]: Another silly question: Isn't `ctx.input` already detached in `forward()`?

**Review by leofang — COMMENTED** [review]
Left two more silly questions 😅 Another note might be useful is to emphasize all of the tensors must live on GPU.

File: docs/source/user_guide/interoperability.rst (hunk 6 above)
Comment (emcastillo) [reply docs/source/user_guide/interoperability.rst:244]: you can't export tensors with `requires_grad` set, so we need to explicitly detach them, otherwise autograd will not work :D

Comment (leofang) [reply docs/source/user_guide/interoperability.rst:244]: Yeah my question is probably more silly that this, let's chat offline!

**Review by leofang — APPROVED** [review]

File: docs/source/user_guide/interoperability.rst [hunk 8]
```diff
@@ -198,6 +198,82 @@ PyTorch also supports zero-copy data exchange through ``DLPack`` (see :ref:`dlpa
    ...     ...
 
 
+Using custom kernels in PyTorch
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (kmaehashi) [inline docs/source/user_guide/interoperability.rst:202]: Seems heading needs to be consistent within a file. :bow:

```suggestion
******************************************
```

File: docs/source/user_guide/interoperability.rst [hunk 9]
```diff
@@ -198,6 +198,82 @@ PyTorch also supports zero-copy data exchange through ``DLPack`` (see :ref:`dlpa
...
+                ((gy_size + bs - 1) // bs,),
+                (cupy_input, cupy_grad_y, cupy_grad_x, gy_size),
+            )
+            # the ownership of the device memory backing cupy_grad_x is implicitly
+            # transferred to torch_y, so this operation is safe even after
+            # going out of scope of this function.
+            torch_grad_x = torch.from_dlpack(cupy_grad_x)
+            return torch_grad_x
+
+
```
Comment (kmaehashi) [inline docs/source/user_guide/interoperability.rst:276]: There're many users still using PyTorch 1.9 or earlier, so maybe worth noting:

```suggestion
.. note::

   Directly feeding a ``torch.Tensor`` to :func:`cupy.from_dlpack` is only supported in the (new) DLPack data exchange protocol added in CuPy v10+ and PyTorch 1.10+.
   For earlier versions, you will need to wrap the ``Tensor`` with ``torch.utils.dlpack.to_dlpack()`` as shown in the above examples.
```

**Review by kmaehashi — DISMISSED** [review]

File: docs/source/user_guide/interoperability.rst [hunk 10]
```diff
@@ -198,6 +198,85 @@ PyTorch also supports zero-copy data exchange through ``DLPack`` (see :ref:`dlpa
...
+            # transferred to torch_y, so this operation is safe even after
+            # going out of scope of this function.
+            torch_grad_x = torch.from_dlpack(cupy_grad_x)
+            return torch_grad_x
+
+.. note::
+
+   Directly feeding a ``torch.Tensor`` to :func:`cupy.from_dlpack` is only supported in the (new) DLPack data exchange protocol added in CuPy v10+ and PyTorch 1.10+.
+   For earlier versions, you will need to wrap the ``Tensor`` with ``torch.utils.dlpack.to_dlpack()`` as shown in the above examples.
 RMM
```
Comment (kmaehashi) [inline docs/source/user_guide/interoperability.rst:280]:
```suggestion

RMM
```

**Review by kmaehashi — APPROVED** [review]

**Comment by kmaehashi** [comment]
pfnCI, test this please.
